### ISSUE TITLE
Add Swift setting to add include paths for a target that are inherited by targets depending on this one

### ISSUE THREAD

**Issue by DougGregor** [swiftlang/swift-package-manager#8077]
### Description

In the [swift-java](https://github.com/swiftlang/swift-java) project, we have a fairly involved operation to figure out the include path for the JNI headers. The only mechanism we have to making this include path available to the Swift source files in that target is via `unsafeFlags`, e.g.,

```swift
    .target(
      name: "JavaRuntime",
      swiftSettings: [
        .unsafeFlags(["-I\(javaIncludePath)", "-I\(javaPlatformIncludePath)"])
      ]
    ),
```

Now, every target that depends JavaRuntime also has to include the same unsafe flags, or importing that module will fail. For example:

```swift
    .target(
      name: "JavaCelsiusConverterGUI",
      dependencies: [ "JavaRuntime" ],
      swiftSettings: [
        // I would rather not need this...
        .unsafeFlags(["-I\(javaIncludePath)", "-I\(javaPlatformIncludePath)"])
      ],
```

Could we have some direct way of specifying C include paths for a SwiftPM target that will affect the C and Swift compiles for that target and any target depending on it? The [`headerSearchPath` C setting](https://developer.apple.com/documentation/packagedescription/csetting/headersearchpath(_:_:)) is almost the right idea, but it requires a relative path, and configuration-derived paths like these need to be absolute (because they are pointing at headers installed by someone else).

### Expected behavior

Expected behavior described above.

### Actual behavior

Actual behavior is "we can't express this".

**Comment by rauhul** [comment]
+1 this could be similar to the includes field of `cc_library` in bazel: https://bazel.build/reference/be/c-cpp#cc_library.includes

**Comment by dschaefer2** [comment]
I am also seeing this when building SwiftPM on Windows. The C defines aren't getting copied over to the Swift modules resulting in symbols from CYaml being exported from consuming modules.

**Comment by rauhul** [comment]
I haven't concretely seen an issue with this, but im worried a bit about Xcc settings with many C modules because theres no way to indicate to the swift compiler that a particular Xcc should only be used when parsing Foo module's headers.

With the current model all Xcc flags are used for every -I which might introduce bug & inconsistent module interfaces.

**Comment by dschaefer2** [comment]
Agreed. Ideally we would do like CMake does and mark settings as public or private. Then you'd only inherit the public ones.

**Comment by DougGregor** [comment]
> Agreed. Ideally we would do like CMake does and mark settings as public or private. Then you'd only inherit the public ones.

There's some potential for footguns here. For example, you may think that your dependency on a C module is private, but the compiler actually needs it in clients to load the module. There's a discussion of the compiler side of this at https://forums.swift.org/t/pitch-allow-non-resilient-modules-to-hide-dependencies-to-clients/75737.

That's not to say we shouldn't separate out public from private, but that folks will be forced to use "public" more than they'd probably like.
